### PR TITLE
fix(core): potential transaction leak in table reader/writer initialization on ZFS

### LINKED ISSUE #4756: Flaky TxnScoreboardTest.testLimitsO3Acquire
### To reproduce

linux-x64-zfs:
```
2024-07-10T09:16:04.4436854Z 2024-07-10T09:16:04.442265Z I i.q.t.TestListener >>>> io.questdb.test.cairo.TxnScoreboardTest.testLimitsO3Acquire
2024-07-10T09:16:04.4443022Z 2024-07-10T09:16:04.443907Z I i.q.t.AbstractTest Starting test TxnScoreboardTest#testLimitsO3Acquire
2024-07-10T09:16:04.4457839Z 2024-07-10T09:16:04.445275Z I i.q.c.TableNameRegistryStore reloading tables file [path=/tmp/junit1858402352682531488/dbRoot/tables.d.0, threadId=11627]
2024-07-10T09:18:17.0195712Z 2024-07-10T09:18:17.015790Z E i.q.t.TestListener ***** Test Failed *****io.questdb.test.cairo.TxnScoreboardTest.testLimitsO3Acquire duration_ms=132573 : 
2024-07-10T09:18:17.0196800Z java.lang.AssertionError: expected:<1> but was:<2>
2024-07-10T09:18:17.0197282Z 	at org.junit.Assert.fail(Assert.java:89)
2024-07-10T09:18:17.0197635Z 	at org.junit.Assert.failNotEquals(Assert.java:835)
2024-07-10T09:18:17.0274984Z 	at org.junit.Assert.assertEquals(Assert.java:647)
2024-07-10T09:18:17.0275689Z 	at org.junit.Assert.assertEquals(Assert.java:633)
2024-07-10T09:18:17.0276181Z 	at io.questdb.test.cairo.TxnScoreboardTest.lambda$testLimits$9(TxnScoreboardTest.java:314)
2024-07-10T09:18:17.0276678Z 	at io.questdb.test.tools.TestUtils.assertMemoryLeak(TestUtils.java:595)
2024-07-10T09:18:17.0277493Z 	at io.questdb.test.cairo.TxnScoreboardTest.testLimits(TxnScoreboardTest.java:278)
2024-07-10T09:18:17.0277962Z 	at io.questdb.test.cairo.TxnScoreboardTest.testLimitsO3Acquire(TxnScoreboardTest.java:337)
2024-07-10T09:18:17.0279474Z 	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2024-07-10T09:18:17.0280062Z 	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2024-07-10T09:18:17.0280534Z 	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2024-07-10T09:18:17.0280970Z 	at java.lang.reflect.Method.invoke(Method.java:566)
2024-07-10T09:18:17.0281354Z 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
2024-07-10T09:18:17.0281768Z 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2024-07-10T09:18:17.0282174Z 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
2024-07-10T09:18:17.0282597Z 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2024-07-10T09:18:17.0283031Z 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
2024-07-10T09:18:17.0283440Z 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
2024-07-10T09:18:17.0284427Z 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
2024-07-10T09:18:17.0284863Z 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
2024-07-10T09:18:17.0285345Z 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
2024-07-10T09:18:17.0285763Z 	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
2024-07-10T09:18:17.0286110Z 	at java.lang.Thread.run(Thread.java:829)
2024-07-10T09:18:17.0286794Z 2024-07-10T09:18:17.015857Z I i.q.t.TestListener <<<< io.questdb.test.cairo.TxnScoreboardTest.testLimitsO3Acquire duration_ms=132573
```

### QuestDB version:

latest master

### OS, in case of Docker specify Docker and the Host OS:

Linux

### File System, in case of Docker specify Host File System:

zfs

### Full Name:

Andrei Pechkurov

### Affiliation:

QuestDB

### Have you followed Linux, MacOs kernel configuration steps to increase Maximum open files and Maximum virtual memory areas limit?

- [X] Yes, I have

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Fixes #4756

It looks like on ZFS the dirty pages may be flushed after we `ftruncate(0)`/`fallocate(2048)`. This manifests as non-zero bytes being read from `TxnScoreboard`'s memory after it's reopen and an exclusive flock was acquired. Adding file zeroing and a `msync()` call to `Files#openCleanRW()` fixes the test flakiness.

Also, `ff.openCleanRW()` call lacked `EINTR` handling, so a `RESTARTABLE` macro is added to the non-blocking `flock()` call.

Other than that, ZFS' `fallocate()` support is basic and `ftruncate()` may hang occasionally:
https://github.com/openzfs/zfs/commit/f734301d2267cbb33eaffbca195fc93f1dae7b74
https://github.com/openzfs/zfs/issues/14290

### CODE REVIEW

File: core/src/test/java/io/questdb/test/cairo/TxnScoreboardTest.java [hunk 1]
```diff
@@ -278,62 +281,64 @@ public void testLimits() throws Exception {
...
-                            Assert.assertEquals(1, scoreboard2.getActiveReaderCount(i + 134));
+                            Assert.assertEquals(1, scoreboard1.getActiveReaderCount(i + 134));
                         } else {
-                            Assert.assertEquals(0, scoreboard2.getActiveReaderCount(i + 134));
+                            Assert.assertEquals(0, scoreboard1.getActiveReaderCount(i + 134));
                         }
                     }
                 }
+
+//                    for (int i = 1; i <= expect; i++) {
```
Comment (ideoma) [inline core/src/test/java/io/questdb/test/cairo/TxnScoreboardTest.java:326]: Delete or uncomment

File: core/src/main/c/share/txn_board.cpp [hunk 2]
```diff
@@ -164,13 +163,15 @@ class txn_scoreboard_t {
     void init(uint32_t entry_count) {
         mask = entry_count - 1;
         size = entry_count;
-        int64_t expected = 0;
-        // Since txn values are guaranteed to be greater than 0, min can be 0 only on
-        // a newly created scoreboard. So, this CAS should only succeed single time.
-        min.compare_exchange_strong(expected, L_MIN);
```
Comment (puzpuzpuz) [inline core/src/main/c/share/txn_board.cpp:170]: TBH I have no idea why we were using CAS operation here considering that this method is called on txn scoreboard initialization, so no one mutates it concurrently.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 18 / 19 (94.74%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|18|19|94.74%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 19 / 19 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TxnScoreboard.java|19|19|100.00%|

**Review by ideoma — DISMISSED** [review]

File: core/src/main/c/share/dedup_comparers.h [hunk 3]
```diff
@@ -24,6 +24,7 @@
 #ifndef QUESTDB_DEDUP_COMPARERS_H
 #define QUESTDB_DEDUP_COMPARERS_H
 
+#include <stddef.h>
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/c/share/dedup_comparers.h:27]: No, the code doesn't compile on my machine without this include:
```
/home/puzpuzpuz/projects/questdb/core/src/main/c/share/dedup_comparers.h:147:38: error: ‘offsetof’ was not declared in this scope
  147 |             auto l_inl = l_val_aux + offsetof(VarcharAuxEntryInlined, chars);
      |
```

File: core/src/test/java/io/questdb/test/FilesTest.java [hunk 4]
```diff
@@ -145,19 +116,23 @@ public void testAllocateConcurrent() throws IOException, InterruptedException {
...
         assertMemoryLeak(() -> {
             File temp = temporaryFolder.newFile();
             TestUtils.writeStringToFile(temp, "abcde");
             try (Path path = new Path().of(temp.getAbsolutePath())) {
                 Assert.assertTrue(Files.exists(path.$()));
-                long fd = Files.openRW(path.$());
                 Assert.assertEquals(5, Files.length(path.$()));
+                long fd = Files.openRW(path.$());
 
+                long M50 = 100 * 1024L * 1024L;
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/test/java/io/questdb/test/FilesTest.java]: Renamed in 47585e25eee6d14fdbe2e60e0eabd9216116db73

**Review by ideoma — APPROVED** [review]

**Comment by puzpuzpuz** [comment]
@ideoma thanks for the review!

**Pull request merged**